File: kevoree-corelibrary/javase/org.kevoree.library.javase.javaseNode/src/main/scala/org/kevoree/library/defaultNodeTypes/jcl/deploy/CommandMapper.scala
package org.kevoree.library.defaultNodeTypes.jcl.deploy

import command._
import org.kevoree.kompare.JavaSePrimitive
import org.kevoree.{Channel, MBinding, Instance, DeployUnit}
import org.kevoree.api.PrimitiveCommand
import org.kevoree.framework.AbstractNodeType
import java.util

/**
 * Created by IntelliJ IDEA.
 * User: duke
 * Date: 30/01/12
 * Time: 17:02
 */

class CommandMapper {

  var nodeType : AbstractNodeType = null
  def setNodeType(n : AbstractNodeType) { nodeType = n }
  def getNodeType : AbstractNodeType = nodeType
  val toClean = new util.ArrayList[EndAwareCommand]

  def doEnd(){
    import scala.collection.JavaConversions._
    toClean.foreach{
      c => c.doEnd
    }
    toClean.clear()
  }

  def buildPrimitiveCommand(p: org.kevoreeAdaptation.AdaptationPrimitive, nodeName: String): PrimitiveCommand = {
    p.getPrimitiveType.getName match {
      case JavaSePrimitive.UpdateDictionaryInstance if(p.getRef.asInstanceOf[Instance].getName == nodeName) => SelfDictionaryUpdate(p.getRef.asInstanceOf[Instance],nodeType)
      case JavaSePrimitive.AddDeployUnit => AddDeployUnit(p.getRef.asInstanceOf[DeployUnit],nodeType.getBootStrapperService)
      case JavaSePrimitive.UpdateDeployUnit => {
        val res = UpdateDeployUnit(p.getRef.asInstanceOf[DeployUnit],nodeType.getBootStrapperService)
        toClean.add(res)
        res
      }
      case JavaSePrimitive.RemoveDeployUnit => {
        val res = RemoveDeployUnit(p.getRef.asInstanceOf[DeployUnit],nodeType.getBootStrapperService)
        toClean.add(res)
        res
      }
      case JavaSePrimitive.AddThirdParty => AddDeployUnit(p.getRef.asInstanceOf[DeployUnit],nodeType.getBootStrapperService)
      case JavaSePrimitive.RemoveThirdParty => RemoveDeployUnit(p.getRef.asInstanceOf[DeployUnit],nodeType.getBootStrapperService)
      case JavaSePrimitive.AddInstance => AddInstance(p.getRef.asInstanceOf[Instance], nodeName, nodeType.getModelService, nodeType.getKevScriptEngineFactory,nodeType.getBootStrapperService)
      case JavaSePrimitive.UpdateDictionaryInstance => UpdateDictionary(p.getRef.asInstanceOf[Instance], nodeName)
      case JavaSePrimitive.RemoveInstance => RemoveInstance(p.getRef.asInstanceOf[Instance], nodeName, nodeType.getModelService, nodeType.getKevScriptEngineFactory,nodeType.getBootStrapperService)
      case JavaSePrimitive.StopInstance  => StartStopInstance(p.getRef.asInstanceOf[Instance], nodeName, false)
      case JavaSePrimitive.StartInstance => StartStopInstance(p.getRef.asInstanceOf[Instance], nodeName, true)
      case JavaSePrimitive.AddBinding => AddBindingCommand(p.getRef.asInstanceOf[MBinding], nodeName)
      case JavaSePrimitive.RemoveBinding => RemoveBindingCommand(p.getRef.asInstanceOf[MBinding], nodeName)
      case JavaSePrimitive.AddFragmentBinding => AddFragmentBindingCommand(p.getRef.asInstanceOf[Channel], p.getTargetNodeName, nodeName)
      case JavaSePrimitive.RemoveFragmentBinding => RemoveFragmentBindingCommand(p.getRef.asInstanceOf[Channel], p.getTargetNodeName, nodeName)
      case JavaSePrimitive.StartThirdParty => NoopCommand()
      case _ => NoopCommand()
    }
  }

}

File: kevoree-corelibrary/javase/org.kevoree.library.javase.javaseNode/src/main/java/org/kevoree/library/defaultNodeTypes/CommandMapper.java
package org.kevoree.library.defaultNodeTypes;

import org.kevoree.Channel;
import org.kevoree.DeployUnit;
import org.kevoree.Instance;
import org.kevoree.MBinding;
import org.kevoree.api.PrimitiveCommand;
import org.kevoree.framework.AbstractNodeType;
import org.kevoree.kompare.JavaSePrimitive;
import org.kevoree.library.defaultNodeTypes.command.*;

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 1/29/13
 * Time: 11:22 AM
 */
public class CommandMapper {

    AbstractNodeType nodeType = null;
    java.util.ArrayList<EndAwareCommand> toClean = new java.util.ArrayList<EndAwareCommand>();

    public void setNodeType(AbstractNodeType n) {
        nodeType = n;
    }

    public void doEnd() {
        for (EndAwareCommand cmd : toClean) {
            cmd.doEnd();
        }
        toClean.clear();
    }

    public PrimitiveCommand buildPrimitiveCommand(org.kevoreeAdaptation.AdaptationPrimitive p, String nodeName) {


        String pTypeName = p.getPrimitiveType().getName();
        if (pTypeName.equals(JavaSePrimitive.UpdateDictionaryInstance())) {
            if (((Instance) p.getRef()).getName().equals(nodeName)) {
                return new SelfDictionaryUpdate((Instance) p.getRef(), nodeType);
            } else {
                return new UpdateDictionary((Instance) p.getRef(), nodeName);
            }

        }

        if (pTypeName.equals(JavaSePrimitive.AddFragmentBinding())) {
            return new AddFragmentBindingCommand((Channel) p.getRef(), p.getTargetNodeName(), nodeName);
        }
        if (pTypeName.equals(JavaSePrimitive.RemoveFragmentBinding())) {
            return new RemoveFragmentBindingCommand((Channel) p.getRef(), p.getTargetNodeName(), nodeName);
        }

        if (pTypeName.equals(JavaSePrimitive.StartInstance())) {
            return new StartStopInstance((Instance) p.getRef(), nodeName, true);
        }
        if (pTypeName.equals(JavaSePrimitive.StopInstance())) {
            return new StartStopInstance((Instance) p.getRef(), nodeName, false);
        }

        if (pTypeName.equals(JavaSePrimitive.AddBinding())) {
            return new AddBindingCommand((MBinding) p.getRef(), nodeName);
        }
        if (pTypeName.equals(JavaSePrimitive.RemoveBinding())) {
            return new RemoveBindingCommand((MBinding) p.getRef(), nodeName);
        }

        if (pTypeName.equals(JavaSePrimitive.AddDeployUnit())) {
            return new AddDeployUnit((DeployUnit) p.getRef(), nodeType.getBootStrapperService());
        }
        if (pTypeName.equals(JavaSePrimitive.RemoveDeployUnit())) {
            RemoveDeployUnit res = new RemoveDeployUnit((DeployUnit) p.getRef(), nodeType.getBootStrapperService());
            toClean.add(res);
            return res;
        }
        if (pTypeName.equals(JavaSePrimitive.UpdateDeployUnit())) {
            UpdateDeployUnit res = new UpdateDeployUnit((DeployUnit) p.getRef(), nodeType.getBootStrapperService());
            toClean.add(res);
            return res;
        }

        if (pTypeName.equals(JavaSePrimitive.AddInstance())) {
            return new AddInstance((Instance) p.getRef(), nodeName, nodeType.getModelService(), nodeType.getKevScriptEngineFactory(), nodeType.getBootStrapperService());
        }
        if (pTypeName.equals(JavaSePrimitive.RemoveInstance())) {
            return new RemoveInstance((Instance) p.getRef(), nodeName, nodeType.getModelService(), nodeType.getKevScriptEngineFactory(), nodeType.getBootStrapperService());
        }

        if (pTypeName.equals(JavaSePrimitive.AddThirdParty())) {
            return new AddDeployUnit((DeployUnit) p.getRef(), nodeType.getBootStrapperService());
        }
        if (pTypeName.equals(JavaSePrimitive.RemoveThirdParty())) {
            return new RemoveDeployUnit((DeployUnit) p.getRef(), nodeType.getBootStrapperService());
        }

        return new NoopCommand();
    }


}
